### PR TITLE
Add warning  message when cloning the HEAD.

### PR DESCRIPTION
**Pull request by ArMour85**
In answer of this [issue](https://github.com/nim-lang/packages/issues/1051), it is important to indicate that the developer use a package without git release tag.

This message, I hope, can also generate more serious in development process and improve the quality. In the common world, a release version has a git tag.

### CODE REVIEW

**Comment by ArMour85** [comment]
The warning message is displayed only if the repo hasn't tags when a version is needed. I made some tests:

**zip** package hasn't tags:
```
./src/nimble install zip@#head
Downloading https://github.com/nim-lang/zip using git
  Verifying dependencies for zip@#head
 Installing zip@#head
   Success: zip installed successfully.

./src/nimble install zip
Downloading https://github.com/nim-lang/zip using git
   Warning: No corresponding tag found with required version. We grab HEAD.
  Verifying dependencies for zip@0.2.1
 Installing zip@0.2.1
    Prompt: zip@0.2.1 already exists. Overwrite? [y/N]
    Answer: y
   Success: zip installed successfully.
```

**karax** package has tags:
```
nimble install karax@#head
Downloading https://github.com/pragmagic/karax using git
   Warning: Package 'karax' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'karax', but file 'button.nim' is in a directory named 'examples' instead. This will be an error in the future.
      Hint: If 'examples' contains source files for building 'karax', rename it to 'karax'. Otherwise, prevent its installation by adding `skipDirs = @["examples"]` to the .nimble file.
  Verifying dependencies for karax@#head
 Installing karax@#head
   Building karax/karax/tools/karun using c backend
   Success: karax installed successfully.

nimble install karax
Downloading https://github.com/pragmagic/karax using git
   Warning: Package 'karax' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'karax', but file 'button.nim' is in a directory named 'examples' instead. This will be an error in the future.
      Hint: If 'examples' contains source files for building 'karax', rename it to 'karax'. Otherwise, prevent its installation by adding `skipDirs = @["examples"]` to the .nimble file.
  Verifying dependencies for karax@1.1.2
 Installing karax@1.1.2
   Building karax/karax/tools/karun using c backend
   Success: karax installed successfully.
```

**Comment by ArMour85** [comment]
I can maybe improve the message.

**Comment by genotrance** [comment]
I'll suggest that message should always be shown even if the user explicitly asks for #head. It should further explain the real problem with #head - if the user installs `pkg1@#head` at some point and then later on `pkg2` requires `pkg1@v0.5.0` which is newer than when `pkg1@#head` was cloned, it won't ever get downloaded by Nimble. `pkg2` will try to build with the old `pkg1@#head` and the build might fail leaving the user guessing what the problem is.

Nimble will assume that `pkg1@#head` in the repo is the latest so it totally throws off dependency version checking. Even if you force install `pkg1@v0.5.0`, I think `#head` will be treated as newer.

Maybe we should always grumble that a package version `#head` is in use and warn the user that it could be out of date. Then of course, the user will want a way to update that to the latest `#head` and the headache never ends.

**Comment by ArMour85** [comment]
@genotrance I totally agree with you to add a warning messages when a user use `#head`. I can add it with the warning I have already added.

**Comment by dom96** [comment]
> Nimble will assume that pkg1@#head in the repo is the latest so it totally throws off dependency version checking. Even if you force install pkg1@v0.5.0, I think #head will be treated as newer.

That's not quite the case I don't think. The distinction here is with how Nim and Nimble handle dependencies. The Nim compiler does not read .nimble files to check versions but Nimble should and does AFAIK.

**Comment by ArMourMag** [comment]
> > Nimble will assume that pkg1@#head in the repo is the latest so it totally throws off dependency version checking. Even if you force install pkg1@v0.5.0, I think #head will be treated as newer.
> 
> That's not quite the case I don't think. The distinction here is with how Nim and Nimble handle dependencies. The Nim compiler does not read .nimble files to check versions but Nimble should and does AFAIK.

Sorry, but I don't understand your answer about the relationship between Nim and Nimble to manage dependencies (I don't very smart). Yes Nim doesn't care about .nimble files but we talk about Nimble when it uses to install, build or run a project. Personally, I rarely run Nim to build my project but rather `nimble build`. In this cases, using `#head` is a problem as @genotrance says.

**Comment by ArMour85** [comment]
> > Nimble will assume that pkg1@#head in the repo is the latest so it totally throws off dependency version checking. Even if you force install pkg1@v0.5.0, I think #head will be treated as newer.
> 
> That's not quite the case I don't think. The distinction here is with how Nim and Nimble handle dependencies. The Nim compiler does not read .nimble files to check versions but Nimble should and does AFAIK.

Sorry, but I don't understand your answer about the relationship between Nim and Nimble to manage dependencies (I don't very smart). Yes Nim doesn't care about .nimble files but we talk about Nimble when it uses to install, build or run a project. Personally, I rarely run Nim to build my project but rather `nimble build`. In this cases, using `#head` is a problem as @genotrance says.

File: src/nimblepkg/download.nim [hunk 1]
```diff
@@ -216,6 +218,9 @@ proc doDownload(url: string, downloadDir: string, verRange: VersionRange,
           display("Switching", "to latest tagged version: " & latest.tag,
                   priority = MediumPriority)
           doCheckout(downMethod, downloadDir, latest.tag)
+      else:
+        display("Warning:", "No corresponding tag found with required version.", Warning, 
+                  priority = HighPriority)
```
Comment (dom96) [inline src/nimblepkg/download.nim:223]: Make the message here the same as above.

File: src/nimblepkg/download.nim [hunk 2]
```diff
@@ -205,6 +205,8 @@ proc doDownload(url: string, downloadDir: string, verRange: VersionRange,
                   onlyTip = not options.forceFullClone)
       else:
         # If no commits have been tagged on the repo we just clone HEAD.
+        display("Warning:", "No corresponding tag found with required version. We grab HEAD.", Warning,
```
Comment (dom96) [inline src/nimblepkg/download.nim:208]:
```suggestion
        display("Warning:", "The package has no tagged releases, downloading HEAD instead.", Warning, 
```

**Review by dom96 — APPROVED** [review]
See remarks, but seems fine otherwise.

**Comment by ArMour85** [comment]
Bump.

Can you merge it please ?

**Comment by dom96** [comment]
Can you make my suggested changes? :)

**Comment by ArMour85** [comment]
Sorry, it's good now.

File: src/nimblepkg/download.nim (hunk 1 above)
Comment (dom96) [reply src/nimblepkg/download.nim]: You forgot this one :)
